src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1215,6 +1224,128 @@ multi_relation_restriction_hook(PlannerInfo *root, RelOptInfo *relOptInfo, Index
 }
 
 
+/*
+ * AdjustReadIntermediateResultCost adjusts the row count and total cost
+ * of a read_intermediate_result call based on the file size.
+ */
+static void
+AdjustReadIntermediateResultCost(RangeTblEntry *rangeTableEntry, RelOptInfo *relOptInfo)
+{
+	PathTarget *reltarget = relOptInfo->reltarget;
+	List *pathList = relOptInfo->pathlist;
+	Path *path = NULL;
+	RangeTblFunction *rangeTableFunction = NULL;
+	FuncExpr *funcExpression = NULL;
+	Const *resultFormatConst = NULL;
+	Datum resultFormatDatum = 0;
+	Oid resultFormatId = InvalidOid;
+	Const *resultIdConst = NULL;
+	Datum resultIdDatum = 0;
+	char *resultId = NULL;
+	int64 resultSize = 0;
+	ListCell *typeCell = NULL;
+	bool binaryFormat = false;
+	double rowCost = 0.;
+	double rowSizeEstimate = 0;
+	double rowCountEstimate = 0.;
+
+	if (rangeTableEntry->rtekind != RTE_FUNCTION ||
+		list_length(rangeTableEntry->functions) != 1)
+	{
+		/* avoid more expensive checks below for non-functions */
+		return;
+	}
+
+	if (!CitusHasBeenLoaded() || !CheckCitusVersion(DEBUG5))
+	{
+		/* read_intermediate_result may not exist */
+		return;
+	}
+
+	if (!ContainsReadIntermediateResultFunction((Node *) rangeTableEntry->functions))
+	{
+		return;
+	}
+
+	rangeTableFunction = (RangeTblFunction *) linitial(rangeTableEntry->functions);
+	funcExpression = (FuncExpr *) rangeTableFunction->funcexpr;
+	resultIdConst = (Const *) linitial(funcExpression->args);
+	if (!IsA(resultIdConst, Const))
+	{
+		/* not sure how to interpret non-const */
+		return;
+	}
+
+	resultIdDatum = resultIdConst->constvalue;
+	resultId = TextDatumGetCString(resultIdDatum);
+
+	resultSize = IntermediateResultSize(resultId);
+	if (resultSize < 0)
+	{
+		/* result does not exist, will probably error out later on */
+		return;
+	}
+
+	resultFormatConst = (Const *) lsecond(funcExpression->args);
+	if (!IsA(resultFormatConst, Const))
+	{
+		/* not sure how to interpret non-const */
+		return;
+	}
+
+	resultFormatDatum = resultFormatConst->constvalue;
+	resultFormatId = DatumGetObjectId(resultFormatDatum);
+
+	if (resultFormatId == BinaryCopyFormatId())
+	{
+		binaryFormat = true;
+
+		/* subtract 11-byte signature + 8 byte header + 2-byte footer */
+		resultSize -= 21;
+	}
+
+	/* postgres' estimate for the width of the rows */
+	rowSizeEstimate += reltarget->width;
+
+	/* add 2 bytes for column count (binary) or line separator (text) */
+	rowSizeEstimate += 2;
+
+	foreach(typeCell, rangeTableFunction->funccoltypes)
+	{
+		Oid columnTypeId = lfirst_oid(typeCell);
+		Oid inputFunctionId = InvalidOid;
+		Oid typeIOParam = InvalidOid;
+
+		if (binaryFormat)
+		{
+			getTypeBinaryInputInfo(columnTypeId, &inputFunctionId, &typeIOParam);
+
+			/* binary format: 4 bytes for field size */
+			rowSizeEstimate += 4;
+		}
+		else
+		{
+			getTypeInputInfo(columnTypeId, &inputFunctionId, &typeIOParam);
+
+			/* text format: 1 byte for tab separator */
+			rowSizeEstimate += 1;
+		}
+
+		rowCost += get_func_cost(inputFunctionId) * cpu_operator_cost;
```

Comment:
Does `read_intermediate_results()` do disk IO? Then maybe add some IO costs? file_fdw adds `seq_page_cost * pages`.

Reply:
If the query that has `read_intermediate_results()` might have `WHERE` clauses, then we might also consider having `baserestrictcost.per_tuple` as part of `rowCost`.